salt/modules/vault.py

```diff
@@ -0,0 +1,164 @@
+# -*- coding: utf-8 -*-
+'''
+:maintainer:    SaltStack
+:maturity:      new
+:platform:      all
+
+Functions to interact with Hashicorp Vault.
+
+:configuration: The salt-master must be configured to allow peer-runner
+    configuration, as well as configuration for the module.
+
+    Add this segment to the master configuration file, or
+    /etc/salt/master.d/vault.conf:
+
+    .. code-block:: yaml
+
+        vault:
+            url: https://vault.service.domain:8200
+            auth:
+                method: token
+                token: 11111111-2222-3333-4444-555555555555
+            policies:
+                - saltstack/minions
+                - saltstack/minion/{minion}
+                .. more policies
+
+    url
+        Url to your Vault installation. Required.
+
+    auth
+        Currently only token auth is supported. The token must be able to create
+        tokens with the policies that should be assigned to minions. Required.
+
+    policies
+        Policies that are assigned to minions when requesting a token. These can
+        either be static, eg saltstack/minions, or templated, eg
+        ``saltstack/minion/{minion}``. ``{minion}`` is shorthand for grains[id].
+        Grains are also available, for example like this:
+        ``my-policies/{grains[os]}``
+
+        Optional. If policies is not configured, saltstack/minions and
+        saltstack/{minion} are used as defaults.
+
+
+    Add this segment to the master configuration file, or
+    /etc/salt/master.d/peer_run.conf:
+
+    .. code-block:: yaml
+
+        peer_run:
+            .*:
+                - vault.generate_token
+'''
+from __future__ import absolute_import
+import logging
+
+import salt.crypt
+import salt.exceptions
+
+
+log = logging.getLogger(__name__)
+
+
+def read_secret(path, key=None):
+    '''
+    Return the value of key at path in vault, or entire secret
+
+    Jinja Example:
+
+    .. code-block:: jinja
+
+            my-secret: {{ salt['vault'].read_secret('secret/my/secret', 'some-key') }}
+
+    .. code-block:: jinja
+
+            {% set supersecret = salt['vault'].read_secret('secret/my/secret') %}
+            secrets:
+                first: {{ supersecret.first }}
+                second: {{ supersecret.second }}
+    '''
+    log.debug('Reading Vault secret for {0} at {1}'.format(__grains__['id'], path))
+    try:
+        url = 'v1/{0}'.format(path)
+        response = __utils__['vault.make_request']('GET', url)
+        if response.status_code != 200:
+            response.raise_for_status()
+        data = response.json()['data']
+
+        if key is not None:
+            return data[key]
+        return data
+    except Exception as e:
+        log.error('Failed to read secret! {0}: {1}'.format(type(e).__name__, e))
+        raise salt.exceptions.CommandExecutionError(e)
+
+
+def write_secret(path, **kwargs):
```

Comment:
The kwargs are actually used to input the key/values to store in Vault, so they aren't actually arguments to the module itself. It just bundles them up into a dict and sends them on. For example, `salt-call vault.write_secret "secret/my/secret" user="foo" password="bar"`. If I'd use `vault.read_secret` on the same path, I'd get back a dict `{'user':'foo', 'password':'bar'}`.

This could be changed to requiring a dict directly, e.g. `salt-call vault.write_secret "secret/my/secret" data='{"user":"foo", "password":"bar"}`, but it makes it somewhat clunkier, IMO.